evalml/pipelines/components/transformers/preprocessing/polynomial_decomposer.py

```diff
@@ -0,0 +1,349 @@
+    """
+
+    name = "Polynomial Decomposer"
+    hyperparameter_ranges = {"degree": Integer(1, 3)}
+    """{
+        "degree": Integer(1, 3)
+    }"""
+    modifies_features = False
+    modifies_target = True
+
+    def __init__(
+        self, time_index: str = None, degree: int = 1, random_seed: int = 0, **kwargs
+    ):
+        if not isinstance(degree, int):
+            if isinstance(degree, float) and degree.is_integer():
+                degree = int(degree)
+            else:
+                raise TypeError(
+                    f"Parameter Degree must be an integer!: Received {type(degree).__name__}",
+                )
+
+        params = {"degree": degree}
+        params.update(kwargs)
+        error_msg = "sktime is not installed. Please install using 'pip install sktime'"
+
+        trend = import_or_raise("sktime.forecasting.trend", error_msg=error_msg)
+        detrend = import_or_raise(
+            "sktime.transformations.series.detrend",
+            error_msg=error_msg,
+        )
+
+        decomposer = detrend.Detrender(trend.PolynomialTrendForecaster(degree=degree))
+
+        params["time_index"] = time_index
+
+        super().__init__(
+            parameters=params,
+            component_obj=decomposer,
+            random_seed=random_seed,
+        )
+
+    def _set_time_index(self, X, y):
+        """Ensures that target data has a pandas.DatetimeIndex that matches feature data."""
+        dt_df = infer_feature_types(X)
+
+        # Use the feature data's index, preferentially
+        if isinstance(dt_df.index, pd.DatetimeIndex):
+            dt_col = pd.Series(dt_df.index)
+        elif dt_df.ww.select("Datetime").shape[1] == 0:
+            raise ValueError(
+                "There are no Datetime features in the feature data and neither the feature or target data doesn't have Datetime index.",
+            )
+        # Use a datetime column of the features if there's only one
+        elif dt_df.ww.select("Datetime").shape[1] == 1:
+            dt_col = dt_df.ww.select("Datetime").squeeze()
+        # With more than one datetime column, use the time_index parameter, if provided.
+        elif dt_df.ww.select("Datetime").shape[1] > 1:
+            if ("time_index" not in self.parameters) or (
+                self.parameters["time_index"] is None
+            ):
+                raise ValueError(
+                    "Too many Datetime features provided in data but no time_index column specified during __init__.",
+                )
+            elif not self.parameters["time_index"] in X:
+                time_index_col = self.parameters["time_index"]
+                raise ValueError(
+                    f"Too many Datetime features provided in data and provided time_index column {time_index_col} not present in data.",
+                )
+            dt_col = dt_df.ww[self.parameters["time_index"]]
+
+        time_index = pd.DatetimeIndex(dt_col, freq=pd.infer_freq(dt_col)).rename(
```

Comment:
I think I have a little bit of an assumption that the data is clean and regularized at this point...maybe that's not a safe assumption?

Reply:
I agree @chukarsten - it should be a safe assumption. We can set an error case here if the the data is irregularly spaced to avoid any issues. WRT to passing the `time_index` - like in the comment above I think we should take the given `time_index` and only fall back to inferring the index when one isn't provided.